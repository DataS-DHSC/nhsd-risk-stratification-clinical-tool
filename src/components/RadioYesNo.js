import React from 'react';
import { Fieldset, Radios } from 'nhsuk-react-components';
import PropTypes from 'prop-types';

const RadioYesNo = ({ field, groupName, prevData, setField }) => {
  const ariaLabel = field.ariaLabel || null;
  return (
    <Fieldset disabled={field.disabled}>
      <Fieldset.Legend
        id={`${field.label.toLowerCase().replace(/ /g, '-')}-radios-label`}
        aria-label={ariaLabel}
      >
        {field.label}
      </Fieldset.Legend>
      <Radios id={groupName} name={groupName} hint={field.hint || ''} inline>
        <Radios.Radio
          key={`${groupName}-yes`}
          id={`${groupName}-yes`}
          value
          defaultChecked={
            !field.disabled && prevData[groupName]
              ? prevData[groupName] === 'true'
              : field.default === 'true'
          }
          required
          onClick={() => setField(groupName, 'true')}
        >
          Yes
        </Radios.Radio>
        <Radios.Radio
          key={`${groupName}-no`}
          id={`${groupName}-no`}
          value={false}
          defaultChecked={
            !field.disabled && prevData[groupName]
              ? prevData[groupName] !== 'true'
              : field.default !== 'true'
          }
          required
          onClick={() => setField(groupName, 'false')}
        >
          No
        </Radios.Radio>
      </Radios>
    </Fieldset>
  );
};

RadioYesNo.propTypes = {
  field: PropTypes.shape().isRequired,
  prevData: PropTypes.shape(),
  groupName: PropTypes.string.isRequired,
  setField: PropTypes.func,
};

RadioYesNo.defaultProps = {
  prevData: {},
  setField: null,
};

export default RadioYesNo;
