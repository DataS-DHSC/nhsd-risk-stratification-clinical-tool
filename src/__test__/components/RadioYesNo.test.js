import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import RadioYesNo from '../../components/RadioYesNo';

const defaultProps = {
  field: {
    label: 'Asthma',
    type: 'radio',
    inline: true,
    default: 'true',
  },
  prevData: {},
  groupName: 'asthma',
  setField: jest.fn(),
};

describe('RadioYesNo', () => {
  it('should take its value from prevData rather than the default', () => {
    const props = { ...defaultProps, prevData: { asthma: 'false' } };
    const container = render(<RadioYesNo {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should take the default value when the field is not in prevData', () => {
    const props = { ...defaultProps, prevData: {} };
    const container = render(<RadioYesNo {...props} />);
    expect(container.asFragment()).toMatchSnapshot();
  });

  it('should call setField on change', () => {
    const setField = jest.fn();
    const props = { ...defaultProps, setField };
    const { container } = render(<RadioYesNo {...props} />);

    userEvent.click(container.querySelector('#asthma-no'));
    expect(setField).toHaveBeenCalledWith('asthma', 'false');

    userEvent.click(container.querySelector('#asthma-yes'));
    expect(setField).toHaveBeenCalledWith('asthma', 'true');
  });
});
