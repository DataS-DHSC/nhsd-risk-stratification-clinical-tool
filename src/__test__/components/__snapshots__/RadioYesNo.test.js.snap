// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioYesNo should take its value from prevData rather than the default 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend"
        id="asthma-radios-label"
      >
        Asthma
      </legend>
      <div
        class="nhsuk-form-group"
      >
        <div
          class="nhsuk-radios nhsuk-radios--inline"
          id="asthma"
          role="radiogroup"
        >
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="asthma-yes--label"
              class="nhsuk-radios__input"
              id="asthma-yes"
              name="asthma"
              required=""
              type="radio"
              value="true"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="asthma-yes"
              id="asthma-yes--label"
            >
              Yes
            </label>
          </div>
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="asthma-no--label"
              checked=""
              class="nhsuk-radios__input"
              id="asthma-no"
              name="asthma"
              required=""
              type="radio"
              value="false"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="asthma-no"
              id="asthma-no--label"
            >
              No
            </label>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;

exports[`RadioYesNo should take the default value when the field is not in prevData 1`] = `
<DocumentFragment>
  <div
    class="nhsuk-form-group"
  >
    <fieldset
      class="nhsuk-fieldset"
    >
      <legend
        class="nhsuk-fieldset__legend"
        id="asthma-radios-label"
      >
        Asthma
      </legend>
      <div
        class="nhsuk-form-group"
      >
        <div
          class="nhsuk-radios nhsuk-radios--inline"
          id="asthma"
          role="radiogroup"
        >
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="asthma-yes--label"
              checked=""
              class="nhsuk-radios__input"
              id="asthma-yes"
              name="asthma"
              required=""
              type="radio"
              value="true"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="asthma-yes"
              id="asthma-yes--label"
            >
              Yes
            </label>
          </div>
          <div
            class="nhsuk-radios__item"
          >
            <input
              aria-labelledby="asthma-no--label"
              class="nhsuk-radios__input"
              id="asthma-no"
              name="asthma"
              required=""
              type="radio"
              value="false"
            />
            <label
              class="nhsuk-label nhsuk-radios__label"
              for="asthma-no"
              id="asthma-no--label"
            >
              No
            </label>
          </div>
        </div>
      </div>
    </fieldset>
  </div>
</DocumentFragment>
`;
